// ory-editor-core/lib/components/Row/inner.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Cell = require('../Cell');

var _Cell2 = _interopRequireDefault(_Cell);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Inner = function Inner(_ref) {
  var _classNames;

  var editable = _ref.editable,
      ancestors = _ref.ancestors,
      _ref$node = _ref.node,
      id = _ref$node.id,
      hover = _ref$node.hover,
      _ref$node$cells = _ref$node.cells,
      cells = _ref$node$cells === undefined ? [] : _ref$node$cells,
      hasInlineChildren = _ref$node.hasInlineChildren,
      containerHeight = _ref.containerHeight,
      blurAllCells = _ref.blurAllCells,
      containerWidth = _ref.containerWidth;
  return _react2.default.createElement(
    'div',
    { className: (0, _classnames2.default)('ory-row', (_classNames = {
        'ory-row-is-hovering-this': Boolean(hover)
      }, _defineProperty(_classNames, 'ory-row-is-hovering-' + (hover || ''), Boolean(hover)), _defineProperty(_classNames, 'ory-row-has-floating-children', hasInlineChildren), _classNames)),
      onClick: blurAllCells
    },
    cells.map(function (c) {
      return _react2.default.createElement(_Cell2.default, {
        rowWidth: containerWidth,
        rowHeight: containerHeight,
        ancestors: [].concat(_toConsumableArray(ancestors), [id]),
        editable: editable,
        key: c,
        id: c
      });
    })
  );
};

exports.default = Inner;
