// ory-editor-core/lib/components/Editable/Inner/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _reselect = require('reselect');

var _classes = require('../../../service/plugin/classes');

var _Cell = require('../../Cell');

var _Cell2 = _interopRequireDefault(_Cell);

var _shouldComponentUpdate = require('../../../helper/shouldComponentUpdate');

var _editable = require('../../../selector/editable');

var _Dimensions = require('../../Dimensions');

var _Dimensions2 = _interopRequireDefault(_Dimensions);

var _cell = require('../../../actions/cell');

var _blur = require('./blur');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Inner = function (_Component) {
  _inherits(Inner, _Component);

  function Inner() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Inner);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Inner.__proto__ || Object.getPrototypeOf(Inner)).call.apply(_ref, [this].concat(args))), _this), _this.shouldComponentUpdate = _shouldComponentUpdate.shouldPureComponentUpdate, _this.createFallbackCell = function () {
      var _this$props = _this.props,
          _this$props$node$cell = _this$props.node.cells,
          cells = _this$props$node$cell === undefined ? [] : _this$props$node$cell,
          createFallbackCell = _this$props.createFallbackCell,
          defaultPlugin = _this$props.defaultPlugin,
          id = _this$props.id;

      if (cells.length === 0) {
        createFallbackCell(new _classes.ContentPlugin(defaultPlugin), id);
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Inner, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      (0, _blur.enableGlobalBlurring)(this.props.blurAllCells);
      this.createFallbackCell();
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      this.createFallbackCell();
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      (0, _blur.disableGlobalBlurring)(this.props.blurAllCells);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          id = _props.id,
          containerWidth = _props.containerWidth,
          containerHeight = _props.containerHeight,
          node = _props.node;

      if (!node) {
        return null;
      }

      var _node$cells = node.cells,
          cells = _node$cells === undefined ? [] : _node$cells;

      return _react2.default.createElement(
        'div',
        { className: 'ory-editable ory-prevent-blur' },
        cells.map(function (c) {
          return _react2.default.createElement(_Cell2.default, {
            rowWidth: containerWidth,
            rowHeight: containerHeight,
            editable: id,
            ancestors: [],
            key: c,
            id: c
          });
        })
      );
    }
  }]);

  return Inner;
}(_react.Component);

var mapStateToProps = (0, _reselect.createStructuredSelector)({ node: _editable.purifiedEditable });

var mapDispatchToProps = { blurAllCells: _cell.blurAllCells, createFallbackCell: _cell.createFallbackCell };

exports.default = (0, _Dimensions2.default)()((0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(Inner));
