// ory-editor-core/lib/reducer/editable/helper/level.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/* eslint no-use-before-define: "off" */
var computeRowLevels = function computeRowLevels(a, b) {
  var _ref = a || {},
      _ref$cells = _ref.cells,
      cells = _ref$cells === undefined ? [] : _ref$cells,
      props = _objectWithoutProperties(_ref, ['cells']);

  var _ref2 = b || {},
      _ref2$left = _ref2.left,
      left = _ref2$left === undefined ? 0 : _ref2$left,
      _ref2$right = _ref2.right,
      right = _ref2$right === undefined ? 0 : _ref2$right,
      _ref2$above = _ref2.above,
      above = _ref2$above === undefined ? 0 : _ref2$above,
      _ref2$below = _ref2.below,
      below = _ref2$below === undefined ? 0 : _ref2$below;

  if (cells.length) {
    props.cells = cells.map(function (c, k) {
      return computeCellLevels(c, {
        left: k === 0 ? left + 1 : 0,
        right: k === cells.length - 1 ? right + 1 : 0,
        above: above + 1,
        below: below + 1
      });
    });
  }

  delete props.levels; // eslint-disable-line prefer-reflect
  return _extends({ levels: { left: left, right: right, above: above, below: below } }, props);
};

var computeCellLevels = function computeCellLevels(a, b) {
  var _ref3 = a || {},
      _ref3$rows = _ref3.rows,
      rows = _ref3$rows === undefined ? [] : _ref3$rows,
      props = _objectWithoutProperties(_ref3, ['rows']);

  var _ref4 = b || {},
      _ref4$left = _ref4.left,
      left = _ref4$left === undefined ? 0 : _ref4$left,
      _ref4$right = _ref4.right,
      right = _ref4$right === undefined ? 0 : _ref4$right,
      _ref4$above = _ref4.above,
      above = _ref4$above === undefined ? 0 : _ref4$above,
      _ref4$below = _ref4.below,
      below = _ref4$below === undefined ? 0 : _ref4$below;

  if (rows.length) {
    props.rows = rows.map(function (r, k) {
      return computeRowLevels(r, {
        left: left + 1,
        right: right + 1,
        above: k === 0 ? above + 1 : 0,
        below: k === rows.length - 1 ? below + 1 : 0
      });
    });
  }

  delete props.levels; // eslint-disable-line prefer-reflect
  return _extends({ levels: { left: left, right: right, above: above, below: below } }, props);
};

var computeDropLevels = exports.computeDropLevels = function computeDropLevels(c) {
  return computeCellLevels(c);
};
